File: hw3-py/task3train.py
from pyspark import SparkContext
import sys
import json
import math
import time
import random
from itertools import combinations


def generate_index(business_list):
    ans = {}
    for i, business in enumerate(business_list):
        ans[business] = i
    return ans


def convert2dict(tuple_list):
    output = {}
    for single_tuple in tuple_list:
        output[single_tuple[0]] = [0,0]
    for single_tuple in tuple_list:
        output[single_tuple[0]][0] += single_tuple[1]
        output[single_tuple[0]][1] += 1
    for key in output.keys():
        output[key] = output[key][0] / output[key][1]
    return output


def pearson_sim(dict1, dict2):
    dict1_set = set(dict1.keys())
    dict2_set = set(dict2.keys())
    common_set = dict1_set.intersection(dict2_set)
    set_len = len(common_set)
    if set_len < 3:
        return -1
    
    r1_mean, r2_mean = 0, 0
    for key in common_set:
        r1_mean += dict1[key]
        r2_mean += dict2[key]
    r1_mean /= set_len
    r2_mean /= set_len
    
    nominator, denominator= 0, 0
    var1, var2 = 0, 0
    for key in common_set:
        nominator += (dict1[key] - r1_mean) * (dict2[key] - r2_mean)
        var1 += (dict1[key] - r1_mean) ** 2
        var2 += (dict2[key] - r2_mean) ** 2
    denominator = math.sqrt(var1) * math.sqrt(var2)
    if denominator == 0:
        return 0
    ans = nominator / denominator
    return ans


def convert_to_file(res_dict, OUTPUT_PATH):

    with open(OUTPUT_PATH, 'w') as fp:
        for single_dict in res_dict:
            fp.write(json.dumps(single_dict))
            fp.write('\n')
        fp.close()
    return


def generate_hash_list(numOfBusiness, numOfHash):
    '''
    return a list that contains a number of hash functions
    a list of [(a, b) , ... ]
    ''' 
    output_list = []
    for _ in range(numOfHash):
        output_list.append((random.randint(200, numOfBusiness), random.randint(200, numOfBusiness)))
    return output_list


def minhash(hash_list, rdd_list, numOfBusiness):
    res = []
    for hash_a, hash_b in hash_list:
        
        global_min = sys.maxsize - 1
        for index in rdd_list:
            local_min = (hash_a * index + hash_b) % numOfBusiness
            if local_min < global_min:
                global_min = local_min

        res.append(global_min)
    return res


def separate_candidate(bids):
    bids = sorted(bids)
    return combinations(bids, 2)


def my_hash(band_list):
    return sum(band_list)


def LSH(minhash_line, numOfBand):
    '''
    
    :return: list([(band_index, band_slice_hash), uid], ... )
    '''
    uid = minhash_line[0]
    # length of minhash
    len_minhash = len(minhash_line[1])
    res = []
    numOfRow = int(len_minhash / numOfBand)
    for i in range(numOfBand):
        band_slice = minhash_line[1][i * numOfRow: (i+1) * numOfRow]
        band_slice_hash = my_hash(band_slice)
        res.append([(i, band_slice_hash), uid])
    return res


def similarity_from_candidate(pair, user_bids_dict):
    set_1, set_2 = set(user_bids_dict[pair[0]]), set(user_bids_dict[pair[1]])
    intersect = set_1.intersection(set_2)
    
    if len(intersect) < 3:
        return -1
    
    union_length = len(set_1) + len(set_2) - len(intersect)
    jaccard = len(intersect) / union_length
    return jaccard


start_time = time.time()

# TASK_PATH = "///Users/tieming/inf553/hw3-py/data/train_review.json"
# OUTPUT_PATH = "///Users/tieming/inf553/hw3-py/task3user.model"
# MODE = "user_based"

TASK_PATH = sys.argv[1]
OUTPUT_PATH = sys.argv[2]
MODE = sys.argv[3]

sc = SparkContext(master="local", appName="task3")


train_review = sc.textFile(TASK_PATH).map(lambda line: json.loads(line))


business_list = train_review.map(lambda line: line["business_id"]).distinct().collect()
user_list = train_review.map(lambda line: line["user_id"]).distinct().collect()
# create dictionaries b_name -> bid, u_name -> uid, bid -> b_name, uid -> u_name
business_to_id = generate_index(business_list)
user_to_id = generate_index(user_list)
id_to_business = dict((v,k) for k,v in business_to_id.items())
id_to_user = dict((v, k) for k, v in user_to_id.items())


if MODE == "item_based":
    item_cf_rdd = train_review.map(lambda line: (line["business_id"], (line["user_id"], line["stars"])))\
            .map(lambda line: (business_to_id[line[0]], (user_to_id[line[1][0]], line[1][1]) ) )\
            .groupByKey()\
            .filter(lambda x: len(x[1]) >= 3)\
            .mapValues(lambda tuple_list: convert2dict(tuple_list))
    # {bid1:{uid1: stars, uid2: stars, ... }, bid2 ....}
    item_cf_dict = item_cf_rdd.collectAsMap()

    # generate all candidate business pairs, with at least 3 co-rated users
    # set numPartition = 2 -> much faster in setting cartesian
    business_rdd = item_cf_rdd.map(lambda line: line[0]).coalesce(2)
    combined_rdd = business_rdd.cartesian(business_rdd)\
            .filter(lambda pair: pair[0] < pair[1])

    res_dict = combined_rdd.map(lambda pair: (pair, pearson_sim(item_cf_dict[pair[0]], item_cf_dict[pair[1]])) )\
            .filter(lambda x: x[1] > 0)\
            .map(lambda pair: {"b1": id_to_business[pair[0][0]], "b2":id_to_business[pair[0][1]], "sim": pair[1]})\
            .collect()

    convert_to_file(res_dict, OUTPUT_PATH)


    
else:

    numOfUser = len(user_list)
    numOfBusiness = len(business_list)
    numOfHash = 30
    numOfBand = 30
    SIMILARITY = 0.01


    user_bids_rdd = train_review.map(lambda line: (user_to_id[line["user_id"]], business_to_id[line["business_id"]]))\
            .groupByKey()\
            .map(lambda line: (line[0], sorted(list(line[1]))))

    user_bids_dict = user_bids_rdd.collectAsMap()

    hash_list = generate_hash_list(numOfBusiness, numOfHash)

    minhash_rdd = user_bids_rdd.map(lambda line: (line[0], minhash(hash_list, line[1], numOfBusiness)))

    lsh_rdd = minhash_rdd.map(lambda line: LSH(line, numOfBand)).flatMap(lambda x: x)\
            .groupByKey()\
            .map(lambda line: (line[0], sorted(list(line[1]))) )\
            .filter(lambda line: len(line[1]) > 1)\
            .flatMapValues(separate_candidate)\
            .map(lambda line: line[1])\
            .distinct()


    filter_by_jaccard = lsh_rdd.map(lambda pair: (pair, similarity_from_candidate(pair, user_bids_dict)))\
            .filter(lambda line: line[1] >= SIMILARITY)\
            .map(lambda x: x[0])

    user_cf_rdd = train_review.map(lambda line: (line["user_id"], (line["business_id"], line["stars"])))\
            .map(lambda line: (user_to_id[line[0]], (business_to_id[line[1][0]], line[1][1]) ) )\
            .groupByKey()\
            .mapValues(lambda tuple_list: convert2dict(tuple_list))

    # {uid1:{bid1: stars, bid2: stars, ... }, uid2 ....}
    user_cf_dict = user_cf_rdd.collectAsMap()

    user_sim = filter_by_jaccard.map(lambda pair: (pair, pearson_sim(user_cf_dict[pair[0]], user_cf_dict[pair[1]])))\
            .filter(lambda x: x[1] > 0)\
            .map(lambda pair: {"u1": id_to_user[pair[0][0]], "u2":id_to_user[pair[0][1]], "sim": pair[1]})\
            .collect()

    convert_to_file(user_sim, OUTPUT_PATH)


print("Duration is: ", time.time() - start_time)

File: hw3-sc/src/main/scala/task3train.scala
import org.apache.spark._
import org.apache.spark.SparkContext._
import org.apache.spark.rdd.RDD
import org.json4s._
import org.apache.log4j._
import org.json4s.jackson.JsonMethods.parse

import scala.collection.mutable.ListBuffer
import scala.collection.mutable._
import scala.collection.Map
import scala.util.Random
import java.io.File
import java.io.PrintWriter

object task3train extends App {

  def set_to_map(user_set: Array[String]): Map[String, Int] = {
    var user_map: Map[String, Int] = Map()
    for (x <- user_set.indices){
      user_map = user_map + (user_set(x) -> x)
    }
    return user_map
  }


  def index_id_convert(user_index: Map[String, Int]): Map[Int, String] ={
    val index_user: Map[Int, String] = user_index.map{
      case (k,v) => (v, k)
    }
    index_user
  }


  def convert_to_map(line:List[(Int, Float)]): Map[Int,Float] = {
    var output_0: Map[Int, MutableList[Float]] = Map()
    for (single <- line){
      output_0 += (single._1 -> MutableList(0,0))
    }
    for (single <- line){
      output_0(single._1)(0) += single._2
      output_0(single._1)(1) += 1
    }
    val output: Map[Int, Float] = output_0.map{
      case (a,MutableList(b,c)) => (a, b/c)
    }
    output
  }


  def pearson_sim(map1: Map[Int, Float], map2: Map[Int, Float]): Double = {
    val set1 = map1.keys.toSet
    val set2 = map2.keys.toSet
    val common_set = set1.intersect(set2)
    if (common_set.size < 3){
      return -1
    }
    var r1_mean: Float = 0
    var r2_mean: Float = 0
    for (key <- common_set){
      r1_mean += map1(key)
      r2_mean += map2(key)
    }
    r1_mean /= common_set.size
    r2_mean /= common_set.size

    var nominator:Double = 0
    var denominator:Double = 0
    var var_1:Double = 0
    var var_2:Double = 0
    for (key <- common_set){
      nominator += (map1(key) - r1_mean) * (map2(key) - r2_mean)
      var_1 += math.pow(map1(key) - r1_mean, 2)
      var_2 += math.pow(map2(key) - r2_mean, 2)
    }
    denominator = math.sqrt(var_1) * math.sqrt(var_2)
    if (denominator == 0){
      return 0
    }
    val ans = nominator / denominator
    ans
  }


  def generate_hash(num_of_business:Long, numOfHash:Int): List[(Int, Int)] = {
    var output_list = new ListBuffer[(Int, Int)]()
    var r = new Random
    r.setSeed(100)

    for(x <- 1 to numOfHash ){
      val curr_tuple: (Int, Int) = (r.nextInt(num_of_business.toInt), r.nextInt(num_of_business.toInt))
      output_list += curr_tuple
    }
    return output_list.toList
  }

  def minHash(hash_list: List[(Int, Int)], rdd_input: List[Int], num_of_business: Long): List[Int] ={
    var res_list = ListBuffer[Int]()
    for (x <- hash_list){
      var global_min: Int = 999999999
      for (i <- rdd_input){
        var local_min: Int = (x._1 * i + x._2) % num_of_business.toInt
        if (local_min < global_min){
          global_min = local_min
        }
      }
      res_list += global_min

    }
    return res_list.toList
  }


  def lsh(minhash_line:(String, List[Int]), num_of_band: Int): List[((Int, Int), String)] ={
    val bid = minhash_line._1
    val minhash_length = minhash_line._2.size
    val num_of_row = minhash_length / num_of_band
    var res = new ListBuffer[((Int, Int), String)]()

    for (i <- 0 until num_of_band){
      var band_slice = minhash_line._2.slice(i * num_of_row, (i+1) * num_of_row)
      var band_hash = band_slice.sum
      res.+=(((i, band_hash), bid))
    }
    return res.toList
  }


  def separate_candidate(bids: List[String]): List[(String, String)] ={
    val combination = (for {i <- bids.indices
                            j <- 0 until i} yield (bids(i), bids(j)))
    return combination.toList
  }


  def convert_to_file(res_map:Array[Map[String,Any]], path:String) = {

    new PrintWriter(new File(path)) {
      res_map.foreach {
        line =>
          write("{" + "\"b1\": " + line("b1") + ", \"b2\": " + line("b2") + ", \"sim\": " + line("sim") + "}")
          write("\n")
      }
      close()
    }
  }

  def similarity_cand(pair:(String,String), user_bid_map:Map[String,List[Int]]): Double = {
    val set_l = user_bid_map(pair._1).toSet
    val set_r = user_bid_map(pair._2).toSet
    val inter = set_l.intersect(set_r)
    if (inter.size < 3){
      return -1
    }
    inter.size.toFloat / (set_l.size + set_r.size - inter.size).toFloat
  }


  /*
   * Scala Main Function
   */

  val t1 = System.nanoTime()
  Logger.getLogger("org").setLevel(Level.ERROR)
  val train_path: String = "///Users/tieming/inf553/hw3-py/data/train_review.json"
  val model_path: String =  "///Users/tieming/inf553/hw3-sc/task3item.model"
  val cf_type: String = "item_based"
  val sc = new SparkContext("local[*]","task3train")

  val rdd_raw: RDD[Map[String, String]] = sc.textFile(train_path).map(line => parse(line).asInstanceOf[Map[String, String]])
  rdd_raw.take(10).foreach(println)
  // generate business and user map ( bid or uid -> index )
  val business_list: Array[String] = rdd_raw.map(line => line("business_id")).distinct().collect()
  val user_list: Array[String] = rdd_raw.map(line => line("user_id")).distinct().collect()

  val user_index: Map[String, Int] = set_to_map(user_list: Array[String])
  val business_index: Map[String, Int] = set_to_map(business_list: Array[String])

  val index_user: Map[Int, String] = index_id_convert(user_index: Map[String, Int])
  val index_business: Map[Int, String] = index_id_convert(business_index: Map[String, Int])

  if (cf_type == "item_based"){
    val item_cf_rdd: RDD[(Int,Map[Int,Float])] = rdd_raw
      .map(line =>(business_index(line("business_id")), (user_index(line("user_id")), line("stars").toFloat)) )
      .groupByKey()
      .mapValues(line => line.toList)
      .filter(x => x._2.size >= 3)
      .mapValues(x => convert_to_map(x))

    val item_cf_map: Map[Int, Map[Int, Float]] = item_cf_rdd.collectAsMap()

    val business_rdd: RDD[Int] = item_cf_rdd.map(line => line._1).coalesce(2)
    val combined_rdd: RDD[(Int, Int)] = business_rdd.cartesian(business_rdd)
      .filter(pair => pair._1 < pair._2)

    val res_map:Array[Map[String,Any]] = combined_rdd.map(line =>(line, pearson_sim(item_cf_map(line._1), item_cf_map(line._2))))
      .filter(x => x._2 > 0)
      .map(line => Map("b1" ->index_business(line._1._1), "b2"->index_business(line._1._2), "sim"->line._2))
      .collect()

    convert_to_file(res_map:Array[Map[String,Any]], model_path:String)
  }

  // user_based model generator
  else{
    val num_of_business: Long = business_list.length
    val num_of_hash: Int = 30
    val num_of_band: Int = 30
    val similarity: Float = 0.01f
    val hash_list: List[(Int, Int)] = generate_hash(num_of_business:Long, num_of_hash:Int)

    val user_bid_rdd:RDD[(String,List[Int])] = rdd_raw
      .map(line => (line("user_id"), line("business_id")) )
      .groupByKey()
      .mapValues(ids => ids.toList)
      .mapValues(ids => ids.map(string => business_index(string)))



    val user_bid_map = user_bid_rdd.collectAsMap()
    val min_hash_rdd = user_bid_rdd
      .mapValues(ids_index => minHash(hash_list, ids_index, num_of_business))

    val lsh_rdd = min_hash_rdd.map(line => lsh(line, num_of_band))
      .flatMap(x=>x)
      .groupByKey()
      .filter(line =>line._2.size > 1)
      .mapValues(line => line.toList)
      .flatMapValues(separate_candidate)
      .map(line => line._2)
      .distinct()

    val filter_by_jaccard: RDD[(Int,Int)] = lsh_rdd.map(line => (line, similarity_cand(line, user_bid_map)))
      .filter(line => line._2 >= similarity)
      .map(x => x._1)
      .map(x => (user_index(x._1), user_index(x._2)))

    val user_cf_rdd: RDD[(Int,Map[Int,Float])] = rdd_raw
      .map(line =>(business_index(line("user_id")), (user_index(line("business_id")), line("stars").trim.toFloat)) )
      .groupByKey()
      .mapValues(line => line.toList)
      .filter(x => x._2.size >= 3)
      .mapValues(x => convert_to_map(x))

    val user_cf_map: Map[Int, Map[Int, Float]] = user_cf_rdd.collectAsMap()

    val res_map:Array[Map[String,Any]] = filter_by_jaccard.map(line =>(line, pearson_sim(user_cf_map(line._1), user_cf_map(line._2))))
      .filter(x => x._2 > 0)
      .map(line => Map("b1" ->index_user(line._1._1), "b2"->index_user(line._1._2), "sim"->line._2))
      .collect()
    convert_to_file(res_map:Array[Map[String,Any]], model_path:String)

  }
  println("Duration: " + ((System.nanoTime() - t1) / 1e9).toString)
}
